### PR TITLE
Swift migration improvements and testing

### PR DESCRIPTION
**Pull request by alazier**

### CODE REVIEW

File: RealmSwift/Migration.swift [hunk 1]
```diff
@@ -102,19 +102,44 @@ public func setSchemaVersion(schemaVersion: UInt, realmPath: String, migrationBl
...
+                      If the Realms at the given path are not encrypted, omit the argument or pass
+                      in `nil`.
 
 :returns: `nil` if the migration was successful, or an `NSError` object that describes the problem
           that occured otherwise.
 */
-public func migrateRealm(path: String) -> NSError? {
-    return RLMRealm.migrateRealmAtPath(path)
+public func migrateRealm(path: String, encryptionKey: NSData? = nil) -> NSError? {
+    if (encryptionKey == nil) {
```
Comment (segiddins) [inline RealmSwift/Migration.swift]: why the parens around the condition?

File: RealmSwift/ObjectSchema.swift [hunk 2]
```diff
@@ -33,7 +33,9 @@ public class ObjectSchema: Equatable {
     var rlmObjectSchema: RLMObjectSchema
 
     /// Array of persisted `Property` objects for an object.
-    public var properties: [Property] { return rlmObjectSchema.properties as [Property] }
+    public var properties: [Property] {
+        return (rlmObjectSchema.properties as [RLMProperty]).map( { Property(rlmProperty: $0) } )
```
Comment (segiddins) [inline RealmSwift/ObjectSchema.swift]: use the same spacing here as on Schema

**Comment by alazier** [comment]
@tgoyne @segiddins 

Only way I was able to make this work was to add configurations to build the dynamic frameworks. I was unable to make xcode behave properly with any other strategies. If you have other suggestions on how to make this work please share. Otherwise this is ready for review.

File: RealmSwift/Tests/MigrationTests.swift [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        }
+        else {
+            migrateRealm(realmPath, encryptionKey: nil)
+        }
+
+        XCTAssertEqual(didRun, shouldRun)
+    }
+
+    // migrate default realm and ensure migration
+    private func migrateAndTestDefaultRealm(shouldRun: Bool = true, schemaVersion: UInt = 1, autoMigration: Bool = false, block: MigrationBlock? = nil) {
```
Comment (segiddins) [inline RealmSwift/Tests/MigrationTests.swift]: why not make this call into `migrateAndTestRealmAtPath` with the `defaultRealmPath`?

Comment (alazier) [inline RealmSwift/Tests/MigrationTests.swift]: this tests the `setDefaultRealmSchemaVersion` method and makes sure `defaultRealm()` performs an automigration

Comment (alazier) [inline RealmSwift/Tests/MigrationTests.swift]: We could still combine the methods though, just not sure it's worth it

Comment (segiddins) [inline RealmSwift/Tests/MigrationTests.swift]: :+1: just as long as it serves a purpose!

Comment (alazier) [inline RealmSwift/Tests/MigrationTests.swift]: I will just remove the default version and test the methods in there separately.

**Comment by alazier** [comment]
@tgoyne @segiddins ready for review

File: Realm/RLMObjectStore.mm [hunk 4]
```diff
@@ -289,18 +289,23 @@ static inline void RLMVerifyInWriteTransaction(RLMRealm *realm) {
...
-                                                                        view:object->_row.get_linklist(prop.column)
-                                                                       realm:object->_realm];
-                if (object.class == NSClassFromString(@"RealmSwift.MigrationObject")) {
-                    [(id<RLMSwiftMigrationObject>)object initalizeListPropertyWithName:prop.name rlmArray:array];
-                }
-                else {
-                    auto list = static_cast<RLMListBase *>(object_getIvar(object, prop.swiftListIvar));
-                    list._rlmArray = array;
+            if (prop.type == RLMPropertyTypeArray) {
+                static Class s_swiftMigrationObjectClass =  NSClassFromString(@"RealmSwift.MigrationObject");
```
Comment (tgoyne) [inline Realm/RLMObjectStore.mm]: extra space after =

File: Realm/RLMObjectStore.mm [hunk 5]
```diff
@@ -289,18 +289,23 @@ static inline void RLMVerifyInWriteTransaction(RLMRealm *realm) {
 
 static inline void RLMInitializeSwiftListAccessor(RLMObjectBase *object) {
     // switch List<> properties to linkviews from standalone arrays
-    if ([object isKindOfClass:NSClassFromString(@"RealmSwift.Object")]) {
+    static Class s_swiftObjectClass = NSClassFromString(@"RealmSwift.Object");
```
Comment (tgoyne) [inline Realm/RLMObjectStore.mm]: This function is very deeply nested and would benefit from some early returns/continues.

File: RealmSwift/Migration.swift [hunk 6]
```diff
@@ -102,19 +102,44 @@ public func setSchemaVersion(schemaVersion: UInt, realmPath: String, migrationBl
 }
 
 /**
+Get the schema version for an encrypted Realm at a given path.
```
Comment (tgoyne) [inline RealmSwift/Migration.swift]: Also non-encrypted since the key is optional?

File: RealmSwift/Migration.swift [hunk 7]
```diff
@@ -176,26 +201,36 @@ public class Migration {
...
-            return listProperties[key]
+        get {
+            if (self.objectSchema[key].type == RLMPropertyType.Array) {
+                return listProperties[key]
+            }
+            return super[key]
+        }
+        set(value) {
+            if (self.objectSchema[key].type == RLMPropertyType.Array) {
+                fatalError("Setting List properties during migrations is unsupported.")
```
Comment (tgoyne) [inline RealmSwift/Migration.swift]: Maybe suggest adding/removing objects from the existing list property?

File: RealmSwift/Tests/TestCase.swift [hunk 8]
```diff
@@ -20,7 +20,7 @@ import XCTest
 import RealmSwift
 import Realm
 
-private func testRealmPath() -> String {
+public func testRealmPath() -> String {
```
Comment (tgoyne) [inline RealmSwift/Tests/TestCase.swift]: internal?

File: RealmSwift/Migration.swift [hunk 9]
```diff
@@ -176,26 +201,36 @@ public class Migration {
...
+            }
+            else {
```
Comment (segiddins) [inline RealmSwift/Migration.swift]: this `else` is redundant

File: RealmSwift/Tests/MigrationTests.swift [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+        // refresh to update realm
+        defaultRealm().refresh()
+
+        // check edited values
+        let object = objects(SwiftObject.self).first!
+        XCTAssertEqual(object.boolCol, false)
+        XCTAssertEqual(object.intCol, 1)
+        XCTAssertEqual(object.floatCol, 1.0 as Float)
+        XCTAssertEqual(object.doubleCol, 10.0)
+        XCTAssertEqual(object.binaryCol, NSData(bytes: "b", length: 0))
```
Comment (segiddins) [inline RealmSwift/Tests/MigrationTests.swift]: that has a `length` of `1`?

File: Realm/RLMObjectStore.mm [hunk 11]
```diff
@@ -289,19 +289,25 @@ static inline void RLMVerifyInWriteTransaction(RLMRealm *realm) {
...
-                }
-                else {
-                    auto list = static_cast<RLMListBase *>(object_getIvar(object, prop.swiftListIvar));
-                    list._rlmArray = array;
-                }
+                [(id<RLMSwiftMigrationObject>)object initalizeListPropertyWithName:prop.name rlmArray:array];
+            }
+            else {
+                auto list = static_cast<RLMListBase *>(object_getIvar(object, prop.swiftListIvar));
+                list._rlmArray = [RLMArrayLinkView arrayWithObjectClassName:prop.objectClassName
```
Comment (segiddins) [inline Realm/RLMObjectStore.mm]: this array creation logic appears to be duplicated?

Comment (alazier) [inline Realm/RLMObjectStore.mm]: I think this is fine. Better than doing the conditional twice as we were before. Not enough code to warrant another helper function.

**Pull request merged**